File: LeetCode/Medium/0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int rowLen = obstacleGrid.size();
        int colLen = obstacleGrid[0].size();
            
        if (obstacleGrid[0][0] == 1 || obstacleGrid[rowLen-1][colLen-1] == 1)
            return 0;
            
        int i = 0;

        while (i < rowLen) {
            if (obstacleGrid[i][0] == 0) obstacleGrid[i++][0] = 1;
            else {
                while (i < rowLen) obstacleGrid[i++][0] = 0;
            }
        }
    
        i = 1;
        
        while (i < colLen){
            if (obstacleGrid[0][i] == 0) obstacleGrid[0][i++] = 1;
            else {
                while (i < colLen) obstacleGrid[0][i++] = 0;
            }
        }

        for (int row = 1; row < rowLen; ++row){
            for (int col = 1; col < colLen; ++col){
                if (obstacleGrid[row][col] != 1)
                    obstacleGrid[row][col] = obstacleGrid[row-1][col] + 
                                                obstacleGrid[row][col-1];
                else
                    obstacleGrid[row][col] = 0;
            }
        }
        
        return obstacleGrid[rowLen-1][colLen-1];
    }
};

File: LeetCode/Medium/0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        if obstacleGrid[0][0] or obstacleGrid[-1][-1]:
            return 0
            
        row_len = len(obstacleGrid)
        col_len = len(obstacleGrid[0])

        for i in range(row_len):
            if obstacleGrid[i][0] == 0:
                obstacleGrid[i][0] = 1
            else:
                while i < row_len:
                    obstacleGrid[i][0] = 0
                    i += 1
                break

        for i in range(1, col_len):
            if obstacleGrid[0][i] == 0:
                obstacleGrid[0][i] = 1
            else:
                while i < col_len:
                    obstacleGrid[0][i] = 0
                    i += 1
                break

        for row in range(1, row_len):
            for col in range(1, col_len):
                if obstacleGrid[row][col] != 1:
                    obstacleGrid[row][col] = obstacleGrid[row-1][col] + \
                                                obstacleGrid[row][col-1]
                else:
                    obstacleGrid[row][col] = 0
        
        return obstacleGrid[row_len-1][col_len-1]